### PR TITLE
redirect to default locale

### LINKED ISSUE #10763: Resolving locales in Shop after a locale being disabled by Admin
**Sylius version affected**: 1.5+

**Description**  
If you use some language in frontend, after disable this language project throw exception:
language code is not define.

**Steps to reproduce**  
1. Create some language in admin area.
2. Go to frontend/public part of store.
3. Change language with last created language
4. Disable this language form channel .
5. Refresh publish/frontend page.

**Possible Solution**  
In my case, just i catch this exception and redirect to main channel language.

Sorry for my bad english.

### PR DESCRIPTION
**Pull request by oallain**
| Q               | A
| --------------- | -----
| Branch?         | 1.5 and after
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes #10763 
| License         | MIT

Redirect to homepage with default locale if locale requested not present in provider.

### CODE REVIEW

**Comment by oallain** [comment]
Hello @lchrusciel ,

I added Behat tests, do you think they are suitable?

File: src/Sylius/Behat/Context/Ui/Shop/LocaleContext.php [hunk 1]
```diff
@@ -16,16 +16,23 @@
...
 
 final class LocaleContext implements Context
 {
     /** @var HomePageInterface */
     private $homePage;
 
-    public function __construct(HomePageInterface $homePage)
-    {
+    /** @var LocaleConverterInterface */
+    private $localeConverter;
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/Shop/LocaleContext.php]: Not needed, please remove

File: src/Sylius/Behat/Context/Ui/Shop/LocaleContext.php [hunk 2]
```diff
@@ -40,6 +47,14 @@ public function iSwitchTheLocaleToTheLocale(string $locale): void
         $this->homePage->switchLocale($locale);
     }
 
+    /**
+     * @When I show homepage with the locale :locale
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/Shop/LocaleContext.php]: We should change this implementation to:
```suggestion
     * @When I show homepage with the locale :localeCode
```
or 
```suggestion
     * @When I show homepage with the locale :language
```

So `\Sylius\Behat\Context\Transform\LocaleContext` will convert value for us.

File: src/Sylius/Bundle/ShopBundle/spec/EventListener/NonChannelLocaleListenerSpec.php [hunk 3]
```diff
@@ -176,10 +185,15 @@ function it_throws_not_found_exception_if_request_locale_is_not_present_in_provi
...
 
-        $this
-            ->shouldThrow(NotFoundHttpException::class)
-            ->during('restrictRequestLocale', [$event])
-        ;
+        $router->generate(
+            'sylius_shop_homepage',
+            ['_locale' => 'ga']
+        )->willReturn('/ga/');
+        $redirect = new RedirectResponse('/ga/');
```
Comment (Zales0123) [inline src/Sylius/Bundle/ShopBundle/spec/EventListener/NonChannelLocaleListenerSpec.php]: I believe we don't need a variable as it's called only once

File: src/Sylius/Bundle/ShopBundle/spec/EventListener/NonChannelLocaleListenerSpec.php [hunk 4]
```diff
@@ -176,10 +185,15 @@ function it_throws_not_found_exception_if_request_locale_is_not_present_in_provi
...
+        $localeProvider->getDefaultLocaleCode()->willReturn('ga');
 
-        $this
-            ->shouldThrow(NotFoundHttpException::class)
-            ->during('restrictRequestLocale', [$event])
-        ;
+        $router->generate(
+            'sylius_shop_homepage',
+            ['_locale' => 'ga']
+        )->willReturn('/ga/');
```
Comment (Zales0123) [inline src/Sylius/Bundle/ShopBundle/spec/EventListener/NonChannelLocaleListenerSpec.php]: Would fit in one line 💃

File: src/Sylius/Bundle/ShopBundle/spec/EventListener/NonChannelLocaleListenerSpec.php [hunk 5]
```diff
@@ -176,10 +185,15 @@ function it_throws_not_found_exception_if_request_locale_is_not_present_in_provi
...
+        $redirect = new RedirectResponse('/ga/');
+
+        $this->restrictRequestLocale($event);
+        $event->setResponse($redirect)->shouldHaveBeenCalledOnce();
```
Comment (Zales0123) [inline src/Sylius/Bundle/ShopBundle/spec/EventListener/NonChannelLocaleListenerSpec.php]: Shouldn't it be placed before a `restrictRequestLocale` call?

Comment (oallain) [reply src/Sylius/Bundle/ShopBundle/spec/EventListener/NonChannelLocaleListenerSpec.php]: After making the change it didn't work.
I did some research and found the documentation that talks about the subject.
And so I think my first version was right.

https://www.phpspec.net/en/latest/manual/prophet-objects.html

File: features/locale/redirect_to_default_locale.feature [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    As a Customer
+    I want to be redirect with default locale if locale is not available
+
+    Background:
+        Given the store operates on a channel named "Web"
+        And that channel allows to shop using "English (United States)" and "French (France)" locales
+        And it uses the "English (United States)" locale by default
+
+    @ui
+    Scenario: Do not Redirect if locale is available
```
Comment (lchrusciel) [inline features/locale/redirect_to_default_locale.feature]:
```suggestion
    Scenario: Stay on the current locale if it is available
```

File: features/locale/redirect_to_default_locale.feature [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        When I browse that channel
+        And I show homepage with the locale "French (France)"
+        Then I should shop using the "français (France)" locale
+
+    @ui
+    Scenario: Redirect to default locale
```
Comment (lchrusciel) [inline features/locale/redirect_to_default_locale.feature]:
```suggestion
    Scenario: Redirect to default locale if it is not available
```

**Comment by pamil** [comment]
The base of this pull-request was changed, you need fetch and reset your local branch 
if you want to add new commits to this pull request. **Reset before you pull, else commits
may become messed-up.**

Unless you added new commits (to this branch) locally that you did not push yet, 
execute `git fetch origin && git reset "redirect-to-default-locale"` to update your local branch.
 
Feel free to ask for assistance when you get stuck :+1:

**Comment by pamil** [comment]
Thanks, @oallain! :1st_place_medal:

**Pull request merged**